File: examples/kotlin/examples/ClientJwtAuthFlow.kt
package io.ekodb.client.examples

import io.ekodb.client.EkoDBClient
import io.github.cdimascio.dotenv.dotenv
import kotlinx.coroutines.runBlocking
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.add
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray
import kotlinx.serialization.json.putJsonObject

/**
 * JWT auth flow — registration + login as pure stored functions.
 *
 * Demonstrates the canonical password-auth pattern using ekoDB stages:
 *
 *   register flow:  BcryptHash → Insert
 *   login flow:     FindOne  → BcryptVerify → If(ok) { JwtSign + Return }
 *   verify flow:    JwtVerify → If(claims != null) { 200 } else { 401 }
 *
 * Operator-owned secrets (the JWT signing key) flow through
 * `{{env.JWT_SECRET}}` so they never enter the function definition or
 * the LLM context window.
 *
 * Requires ekoDB >= 0.43.0.
 */
fun main() = runBlocking {
    val dotenv = dotenv { ignoreIfMissing = true }
    val baseUrl = dotenv["API_BASE_URL"] ?: "http://localhost:8080"
    val apiKey = dotenv["API_BASE_KEY"] ?: "a-test-api-key-from-ekodb"

    val client = EkoDBClient.builder()
        .baseUrl(baseUrl)
        .apiKey(apiKey)
        .build()

    println("✓ Client created")

    // 1. Register: bcrypt-hash, insert.
    val register = buildJsonObject {
        put("label", "kt_users_register")
        put("name", "Register user")
        put("description", "Validate, bcrypt-hash, insert.")
        putJsonObject("parameters") {
            putJsonObject("email") { put("required", true) }
            putJsonObject("password") { put("required", true) }
        }
        putJsonArray("functions") {
            add(buildJsonObject {
                put("type", "BcryptHash")
                put("plain", "{{password}}")
                put("cost", 12)
                put("output_field", "password_hash")
            })
            add(buildJsonObject {
                put("type", "Insert")
                put("collection", "kt_users")
                putJsonObject("record") {
                    put("email", "{{email}}")
                    put("password_hash", "{{password_hash}}")
                }
                put("bypass_ripple", false)
            })
        }
    }
    saveJwtAuthFn(client, register, "kt_users_register")

    // 2. Login: find user, verify bcrypt, sign JWT on success.
    val login = buildJsonObject {
        put("label", "kt_users_login")
        put("name", "Login user")
        put("description", "Verify password, mint JWT.")
        putJsonObject("parameters") {
            putJsonObject("email") { put("required", true) }
            putJsonObject("password") { put("required", true) }
        }
        putJsonArray("functions") {
            add(buildJsonObject {
                put("type", "FindOne")
                put("collection", "kt_users")
                put("key", "email")
                put("value", "{{email}}")
            })
            add(buildJsonObject {
                put("type", "BcryptVerify")
                put("plain", "{{password}}")
                put("hash_field", "password_hash")
                put("output_field", "password_ok")
            })
            add(buildJsonObject {
                put("type", "If")
                putJsonObject("condition") {
                    put("type", "FieldEquals")
                    putJsonObject("value") {
                        put("field", "password_ok")
                        put("value", true)
                    }
                }
                putJsonArray("then_functions") {
                    add(buildJsonObject {
                        put("type", "JwtSign")
                        putJsonObject("claims") { put("email", "{{email}}") }
                        put("secret", "{{env.JWT_SECRET}}")
                        put("algorithm", "HS256")
                        put("expires_in_secs", 3600)
                        put("output_field", "token")
                    })
                    add(buildJsonObject {
                        put("type", "Return")
                        putJsonObject("fields") {
                            put("ok", true)
                            put("token", "{{token}}")
                        }
                        put("status_code", 200)
                    })
                }
                putJsonArray("else_functions") {
                    add(buildJsonObject {
                        put("type", "Return")
                        putJsonObject("fields") { put("ok", false) }
                        put("status_code", 401)
                    })
                }
            })
        }
    }
    saveJwtAuthFn(client, login, "kt_users_login")

    // 3. Verify a JWT — fail-closed when claims is null.
    val verify = buildJsonObject {
        put("label", "kt_users_verify_token")
        put("name", "Verify JWT token")
        putJsonObject("parameters") {
            putJsonObject("token") { put("required", true) }
        }
        putJsonArray("functions") {
            // Synthetic record so JwtVerify has working_data[0] to read off.
            add(buildJsonObject {
                put("type", "Insert")
                put("collection", "_inflight_jwt_check")
                putJsonObject("record") { put("token", "{{token}}") }
                put("bypass_ripple", true)
                put("ttl", 60)
            })
            add(buildJsonObject {
                put("type", "JwtVerify")
                put("token_field", "token")
                put("secret", "{{env.JWT_SECRET}}")
                put("algorithm", "HS256")
                put("output_field", "claims")
            })
            add(buildJsonObject {
                put("type", "If")
                putJsonObject("condition") {
                    put("type", "FieldEquals")
                    putJsonObject("value") {
                        put("field", "claims")
                        // null literal in JsonObject form
                        put("value", JsonPrimitive(null as String?))
                    }
                }
                putJsonArray("then_functions") {
                    add(buildJsonObject {
                        put("type", "Return")
                        putJsonObject("fields") { put("ok", false) }
                        put("status_code", 401)
                    })
                }
                putJsonArray("else_functions") {
                    add(buildJsonObject {
                        put("type", "Return")
                        putJsonObject("fields") { put("ok", true) }
                        put("status_code", 200)
                    })
                }
            })
        }
    }
    saveJwtAuthFn(client, verify, "kt_users_verify_token")

    println("\n=== Auth flow defined as pure stored functions ===")
    println("Call them like:")
    println("  POST /api/functions/kt_users_register { \"email\": \"a@b.com\", \"password\": \"s3cret\" }")
    println("  POST /api/functions/kt_users_login    { \"email\": \"a@b.com\", \"password\": \"s3cret\" }")
    println("  POST /api/functions/kt_users_verify_token { \"token\": \"<jwt>\" }")

    for (label in listOf("kt_users_register", "kt_users_login", "kt_users_verify_token")) {
        try {
            client.deleteUserFunction(label)
        } catch (_: Exception) {
        }
    }
    println("\n✓ Cleaned up demo functions")
}

private suspend fun saveJwtAuthFn(
    client: EkoDBClient,
    fn: kotlinx.serialization.json.JsonObject,
    label: String,
) {
    try {
        client.saveUserFunction(fn)
        println("✓ $label saved")
    } catch (e: Exception) {
        println("SaveUserFunction($label) error: ${e.message}")
    }
}

File: examples/rust/examples/client_jwt_auth_flow.rs
//! JWT auth flow — registration + login as pure stored functions.
//!
//! Demonstrates the canonical password-auth pattern using ekoDB stages
//! end-to-end:
//!
//!   register flow:  Validate → BcryptHash → Insert
//!   login flow:     FindOne  → BcryptVerify → If(ok) { JwtSign + Return }
//!   verify flow:    JwtVerify → If(claims != null) { Return ok } else { Return 401 }
//!
//! Nothing in the client needs to host bcrypt or JWT primitives — the app
//! sends the password through a `{{password}}` placeholder and ekoDB
//! handles hashing, verification, and signing. Operator-owned secrets
//! (the JWT signing key) flow through `{{env.JWT_SECRET}}` so they never
//! enter the LLM context window or the function definition stored on
//! disk.
//!
//! Requires ekoDB >= 0.43.0.

use ekodb_client::{Client, Function, FunctionCondition, ParameterDefinition, UserFunction};
use serde_json::json;
use std::collections::HashMap;
use std::env;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    dotenv::dotenv().ok();

    let base_url = env::var("API_BASE_URL").unwrap_or_else(|_| "http://localhost:8080".to_string());
    let api_key =
        env::var("API_BASE_KEY").unwrap_or_else(|_| "a-test-api-key-from-ekodb".to_string());

    let client = Client::builder()
        .base_url(&base_url)
        .api_key(&api_key)
        .build()?;

    println!("✓ Client created");

    // Stage 1: register — bcrypt-hash, insert.
    let register = UserFunction::new("rs_users_register", "Register user")
        .with_description("Validate, bcrypt-hash, insert.")
        .with_parameter(ParameterDefinition::new("email").required())
        .with_parameter(ParameterDefinition::new("password").required())
        .with_function(Function::BcryptHash {
            plain: "{{password}}".to_string(),
            cost: Some(12),
            output_field: "password_hash".to_string(),
        })
        .with_function(Function::Insert {
            collection: "rs_users".to_string(),
            record: json!({
                "email": "{{email}}",
                "password_hash": "{{password_hash}}",
            }),
            bypass_ripple: None,
            ttl: None,
        });
    let _ = client.save_user_function(register).await;
    println!("✓ rs_users_register saved");

    // Stage 2: login — find the user, verify bcrypt, sign a JWT on success.
    let mut claims = HashMap::new();
    claims.insert(
        "email".to_string(),
        serde_json::Value::String("{{email}}".to_string()),
    );

    let login = UserFunction::new("rs_users_login", "Login user")
        .with_description("Verify password, mint JWT.")
        .with_parameter(ParameterDefinition::new("email").required())
        .with_parameter(ParameterDefinition::new("password").required())
        .with_function(Function::FindOne {
            collection: "rs_users".to_string(),
            key: "email".to_string(),
            value: serde_json::Value::String("{{email}}".to_string()),
        })
        .with_function(Function::BcryptVerify {
            plain: "{{password}}".to_string(),
            hash_field: "password_hash".to_string(),
            output_field: "password_ok".to_string(),
        })
        .with_function(Function::If {
            condition: FunctionCondition::FieldEquals {
                field: "password_ok".to_string(),
                value: serde_json::Value::Bool(true),
            },
            then_functions: vec![
                Box::new(Function::JwtSign {
                    claims,
                    secret: "{{env.JWT_SECRET}}".to_string(),
                    algorithm: Some("HS256".to_string()),
                    expires_in_secs: Some(3600),
                    output_field: "token".to_string(),
                }),
                Box::new(Function::Return {
                    fields: HashMap::from([
                        ("ok".to_string(), serde_json::Value::Bool(true)),
                        (
                            "token".to_string(),
                            serde_json::Value::String("{{token}}".to_string()),
                        ),
                    ]),
                    status_code: Some(200),
                }),
            ],
            else_functions: Some(vec![Box::new(Function::Return {
                fields: HashMap::from([("ok".to_string(), serde_json::Value::Bool(false))]),
                status_code: Some(401),
            })]),
        });
    let _ = client.save_user_function(login).await;
    println!("✓ rs_users_login saved");

    // Stage 3: verify — read the token off the request, decode + validate
    // the signature, branch on null claims.
    let verify = UserFunction::new("rs_users_verify_token", "Verify JWT token")
        .with_description("Decode + validate JWT; null claims = reject.")
        .with_parameter(ParameterDefinition::new("token").required())
        .with_function(Function::Insert {
            // Synthetic record so JwtVerify has a `working_data[0]` to read
            // the token field off. (In a real flow the token would arrive
            // on the request path or an existing record.)
            collection: "_inflight_jwt_check".to_string(),
            record: json!({ "token": "{{token}}" }),
            bypass_ripple: Some(true),
            ttl: Some(serde_json::json!(60)),
        })
        .with_function(Function::JwtVerify {
            token_field: "token".to_string(),
            secret: "{{env.JWT_SECRET}}".to_string(),
            algorithm: Some("HS256".to_string()),
            output_field: "claims".to_string(),
        })
        .with_function(Function::If {
            condition: FunctionCondition::FieldEquals {
                field: "claims".to_string(),
                value: serde_json::Value::Null,
            },
            then_functions: vec![Box::new(Function::Return {
                fields: HashMap::from([("ok".to_string(), serde_json::Value::Bool(false))]),
                status_code: Some(401),
            })],
            else_functions: Some(vec![Box::new(Function::Return {
                fields: HashMap::from([("ok".to_string(), serde_json::Value::Bool(true))]),
                status_code: Some(200),
            })]),
        });
    let _ = client.save_user_function(verify).await;
    println!("✓ rs_users_verify_token saved");

    println!("\n=== Auth flow defined as pure stored functions ===");
    println!("Call them like:");
    println!("  POST /api/functions/rs_users_register {{ \"email\": \"a@b.com\", \"password\": \"s3cret\" }}");
    println!("  POST /api/functions/rs_users_login    {{ \"email\": \"a@b.com\", \"password\": \"s3cret\" }}");
    println!("  POST /api/functions/rs_users_verify_token {{ \"token\": \"<jwt>\" }}");
    println!(
        "\nSet JWT_SECRET in ekoDB's environment_vars whitelist before invoking — \
         the {{{{env.JWT_SECRET}}}} placeholder reads from that whitelist, NEVER from \
         the function definition."
    );

    // Cleanup so re-runs don't conflict with the first registration.
    let _ = client.delete_user_function("rs_users_register").await;
    let _ = client.delete_user_function("rs_users_login").await;
    let _ = client.delete_user_function("rs_users_verify_token").await;
    println!("\n✓ Cleaned up demo functions");

    Ok(())
}
